### PR TITLE
Support Client Feature flags

### PR DESCRIPTION
**Pull request by demirkayaender**
**What changed?**
Shipping WorkflowExecutionAlreadyCompletedError broke some customers' workflow logic since it was changing previous contract between their workflow and cadence server. With this change, they can enable the new feature only when they are ready. 

Changes need to go in with the following orders
- IDL changes: https://github.com/uber/cadence-idl/pull/73
- Client-side changes: https://github.com/uber-go/cadence-client/pull/1103
- Server-side changes: https://github.com/uber/cadence/pull/4257
- Samples changes: https://github.com/uber-common/cadence-samples/pull/45

**Why?**
Without this change, shipping new features will break customer workflows.

**How did you test it?**
`go test -run TestIntegrationSuite github.com/uber/cadence/host -testify.m TestSignalWorkflow$ -v`

CLI + helloworkflow and cron workflow

**Potential risks**

### CODE REVIEW

File: client/client.go [hunk 1]
```diff
@@ -47,6 +47,9 @@ type (
 	// Options are optional parameters for Client creation.
 	Options = internal.ClientOptions
 
+	// FeatureFlags define which breaking changes can be enabled for client
+	FeatureFlags = internal.FeatureFlags
+
```
Comment (Groxx) [inline client/client.go:52]: this diff is fairly large already, but tbh I wonder if it's worth saying "nope, we should stop doing this immediately", and doing the new type + cast-or-whatever.  IDEs and code-generators really don't play well with these internal-aliases...

File: internal/internal_utils.go [hunk 2]
```diff
@@ -64,15 +67,53 @@ const (
...
-	yarpcCallOptions = []yarpc.CallOption{
+	_yarpcCallOptions = []yarpc.CallOption{
 		yarpc.WithHeader(libraryVersionHeaderName, LibraryVersion),
 		yarpc.WithHeader(featureVersionHeaderName, FeatureVersion),
 		yarpc.WithHeader(clientImplHeaderName, clientImplHeaderValue),
 	}
 )
 
+func fromInternalFeatureFlags(featureFlags FeatureFlags) s.FeatureFlags {
+	return s.FeatureFlags{
```
Comment (demirkayaender) [inline internal/internal_utils.go:86]: Remove client only header here

File: internal/internal_workflow_client.go [hunk 3]
```diff
@@ -1026,7 +1028,9 @@ func (dc *domainClient) Describe(ctx context.Context, name string) (*s.DescribeD
 	var response *s.DescribeDomainResponse
 	err := backoff.Retry(ctx,
 		func() error {
-			tchCtx, cancel, opt := newChannelContext(ctx)
+
+			// TODO change here, opt should have the feature flags
```
Comment (demirkayaender) [inline internal/internal_workflow_client.go:1032]: remove here

**Review by demirkayaender — COMMENTED** [review]
Talked to Emrah and added offline comments as PR comment, will address them

**Review by emrahs — APPROVED** [review]

**Pull request merged**